### PR TITLE
add part 09 step 24

### LINKED ISSUE #373: Write Part - 9 Step - 24 & unit test
Implement Step24Controller and write all unit tests. Verify the display name of the test conforms to the pattern of "Part X Step Y". Follow the standardize naming, comments, etc. to be consistent.

### PR DESCRIPTION
**Pull request by mmschaefer**
Fix #373 - add part09 step24 and associated unit tests.

### CODE REVIEW

File: src/org/etools/j1939_84/controllers/part09/Part09Step24Controller.java [hunk 1]
```diff
@@ -54,12 +57,52 @@
...
+            // for any with non- zero timer values created earlier in step 6.9.7.1 and fail if any timer value is less
+            // than the value it was earlier in this part.
+            for (EngineHoursTimer previousTimer : previousDM33Timers) {
+                if (previousTimer.getEiAecdTimer1() > 0 || previousTimer.getEiAecdTimer2() > 0) {
+                    int timeId = previousTimer.getEiAecdNumber();
+                    var currentTimer = packet.getTimer(timeId);
+                    if (currentTimer != null) {
+                        if (currentTimer.getEiAecdTimer1() < previousTimer.getEiAecdTimer1()) {
+                            addFailure("6.9.24.2.b - ECU " + packet.getModuleName()
+                                    + " reported 6.9.7.1 timer 1 value less than previously observed in 6.9.7.1");
```
Comment (gumbelmj) [inline src/org/etools/j1939_84/controllers/part09/Part09Step24Controller.java]: this needs to be:
```                                    + " reported timer 1 value less than previously observed in 6.9.7.1");```

File: src/org/etools/j1939_84/controllers/part09/Part09Step24Controller.java [hunk 2]
```diff
@@ -54,12 +57,52 @@
...
+                        }
+                        if (currentTimer.getEiAecdTimer2() < previousTimer.getEiAecdTimer2()) {
+                            addFailure("6.9.24.2.b - ECU " + packet.getModuleName()
+                                    + " reported 6.9.7.1 timer 2 value less than previously observed in 6.9.7.1");
```
Comment (gumbelmj) [inline src/org/etools/j1939_84/controllers/part09/Part09Step24Controller.java]: this needs to be:
```+ " reported timer 2 value less than previously observed in 6.9.7.1");```

File: src/org/etools/j1939_84/controllers/part09/Part09Step24Controller.java [hunk 3]
```diff
@@ -54,12 +57,52 @@
...
+                        }
+
+                    }
+                }
+            }
+            if (!packetTimers.containsAll(previousDM33Timers)) {
+                addFailure("6.9.24.2.b - ECU " + packet.getModuleName()
```
Comment (gumbelmj) [inline src/org/etools/j1939_84/controllers/part09/Part09Step24Controller.java]: This failure was not asked for and needs to be removed.

File: src/org/etools/j1939_84/controllers/part09/Part09Step24Controller.java [hunk 4]
```diff
@@ -54,12 +57,52 @@
     @Override
     protected void run() throws Throwable {
         // 6.9.24.1.a. DS DM33 [(send Request (PGN 59904) for PGN 41216 (SPNs 4124-4126))] to each OBD ECU.
-        // 6.9.24.2.a. Fail if any ECU reports a different number EI-AECD than was reported in part 2.
-        // [Engines using SI technology need not respond until the 2024 engine model year].
```
Comment (gumbelmj) [inline src/org/etools/j1939_84/controllers/part09/Part09Step24Controller.java]: The check needs to take into consideration the fuel type and model year.
The logic works out to be :
```
if (!sparkIgnition || getEngineModelYear() >= 2024) {
// do testing
}
```

**Review by gumbelmj — CHANGES_REQUESTED** [review]
Changes are needed:
- working on failures needs update
- SI engines needs factored in
- additional failure needs removed.

File: src/org/etools/j1939_84/controllers/part09/Part09Step24Controller.java (hunk 4 above)
Comment (mmschaefer) [reply src/org/etools/j1939_84/controllers/part09/Part09Step24Controller.java]: Does not say anything about that.

Comment (mmschaefer) [reply src/org/etools/j1939_84/controllers/part09/Part09Step24Controller.java]: Requirements are: 

6.9.24.2 Fail Criteria

a. Fail if any ECU reports a different number EI-AECD than was reported in part 2. [Engines using SI technology need not respond until the 2024 engine model year].

b. Compare to list of ECU address + EI-AECD number + actual time (for Timer 1 and/or Timer 2) for any with nonzero timer values created earlier in step 6.9.6.1 and fail if any timer value is less than the value it was earlier in this part.

c. Fail if NACK not received from OBD ECUs that did not provide a DM33 message.

Comment (mmschaefer) [reply src/org/etools/j1939_84/controllers/part09/Part09Step24Controller.java]: I do not see what you mean.

Comment (gumbelmj) [reply src/org/etools/j1939_84/controllers/part09/Part09Step24Controller.java]: For point C, we don't fail the engine is using SI and the model year is < 2024.

**Review by gumbelmj — APPROVED** [review]

**Pull request merged**